presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/StringSqlFunctions.java

```diff
@@ -32,4 +32,13 @@ public static String replaceFirst()
     {
         return "RETURN IF(replace IS NULL, NULL, IF(STRPOS(str, search) = 0, str, SUBSTR(str, 1, STRPOS(str, search) - 1) || replace || SUBSTR(str, STRPOS(str, search) + LENGTH(search))))";
     }
+
+    @SqlInvokedScalarFunction(value = "trail", deterministic = true, calledOnNullInput = true)
+    @Description("Returns the last N characters of the input string.")
+    @SqlParameters({@SqlParameter(name = "str", type = "varchar"), @SqlParameter(name = "N", type = "int")})
+    @SqlType("varchar")
+    public static String trail()
+    {
+        return "RETURN REVERSE(SUBSTR(REVERSE(str), 1, N))";
```

Comment:
Maybe just:

return substr(str, greatest(1, length(str)-N+1), N)

reverse twice creates two objects. So this version will be more efficient

Reply:
Also you need to check N >= 0

Reply:
Did some local testing.

substr(str, greatest(1, length(str)-N+1), N)
<img width="344" alt="Screenshot 2023-08-30 at 9 36 47 AM" src="https://github.com/prestodb/presto/assets/13852634/132a885f-9dee-48fa-9203-6ec1e78331a5">

REVERSE(SUBSTR(REVERSE(str), 1, N))
<img width="450" alt="Screenshot 2023-08-30 at 9 37 47 AM" src="https://github.com/prestodb/presto/assets/13852634/0346090e-1ba6-492d-8437-93b48b5b3541">


REVERSE(SUBSTR(REVERSE(str), 1, N)) appears to consume fewer resources.

Reply:
Added test cases for N >= 0

Reply:
Hmm can you add a SqlBenchMark class so we can be absolutely sure? This is counter intuitve.

Reply:
Results:

RETURN REVERSE(SUBSTR(REVERSE(str), 1, N)):
sql_trail ::  105.311 
cpu ms ::    0B 
peak memory :: 
in   15K,      0B,    142K/s,      0B/s :: 
out   15K,   103KB,    142K/s,   973KB/s

Reply:
Results:

RETURN substr(str, greatest(1, length(str)-N+1), N):
TBD

Reply:
Hmm ok. Can you just copy paste the results from the benchmark run here? The ops/ns is the most intersteing measure

Reply:
peak_memory:0,elapsed_millis:102,input_rows_per_second:147214,output_rows_per_second:147214,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:101892220,cpu_nanos:101037000,user_nanos:100607000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291464
peak_memory:0,elapsed_millis:100,input_rows_per_second:150405,output_rows_per_second:150405,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:99730230,cpu_nanos:98942000,user_nanos:98580000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290352
peak_memory:0,elapsed_millis:106,input_rows_per_second:141785,output_rows_per_second:141785,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:105793824,cpu_nanos:100322000,user_nanos:99727000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291393
peak_memory:0,elapsed_millis:98,input_rows_per_second:153638,output_rows_per_second:153638,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:97631751,cpu_nanos:97374000,user_nanos:96867000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291381
peak_memory:0,elapsed_millis:103,input_rows_per_second:145496,output_rows_per_second:145496,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:103095413,cpu_nanos:101957000,user_nanos:100979000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290982
peak_memory:0,elapsed_millis:103,input_rows_per_second:145259,output_rows_per_second:145259,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:103263417,cpu_nanos:102239000,user_nanos:101665000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291521
peak_memory:0,elapsed_millis:95,input_rows_per_second:158205,output_rows_per_second:158205,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:94813396,cpu_nanos:94572000,user_nanos:94355000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290834
peak_memory:0,elapsed_millis:112,input_rows_per_second:134173,output_rows_per_second:134173,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:111795588,cpu_nanos:111013000,user_nanos:104894000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:292319
peak_memory:0,elapsed_millis:180,input_rows_per_second:83473,output_rows_per_second:83473,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:179697906,cpu_nanos:178978000,user_nanos:142083000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:293188
peak_memory:0,elapsed_millis:169,input_rows_per_second:88582,output_rows_per_second:88582,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:169334531,cpu_nanos:167618000,user_nanos:134220000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:292495
peak_memory:0,elapsed_millis:101,input_rows_per_second:148870,output_rows_per_second:148870,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:100758636,cpu_nanos:95394000,user_nanos:93757000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291171
peak_memory:0,elapsed_millis:111,input_rows_per_second:135554,output_rows_per_second:135554,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:110656508,cpu_nanos:109364000,user_nanos:108610000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290851
peak_memory:0,elapsed_millis:96,input_rows_per_second:156499,output_rows_per_second:156499,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:95847013,cpu_nanos:95747000,user_nanos:95566000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290947
peak_memory:0,elapsed_millis:99,input_rows_per_second:151814,output_rows_per_second:151814,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:98804537,cpu_nanos:95724000,user_nanos:95229000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290857
peak_memory:0,elapsed_millis:105,input_rows_per_second:143215,output_rows_per_second:143215,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:104737108,cpu_nanos:104698000,user_nanos:104518000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290017
peak_memory:0,elapsed_millis:115,input_rows_per_second:130707,output_rows_per_second:130707,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:114759775,cpu_nanos:112704000,user_nanos:111856000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291624
peak_memory:0,elapsed_millis:110,input_rows_per_second:136553,output_rows_per_second:136553,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:109847244,cpu_nanos:109811000,user_nanos:109501000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:292182
peak_memory:0,elapsed_millis:106,input_rows_per_second:141739,output_rows_per_second:141739,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:105827789,cpu_nanos:105361000,user_nanos:104797000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290892
peak_memory:0,elapsed_millis:115,input_rows_per_second:130695,output_rows_per_second:130695,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:114770647,cpu_nanos:107582000,user_nanos:99920000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290964
peak_memory:0,elapsed_millis:90,input_rows_per_second:167496,output_rows_per_second:167496,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:89554270,cpu_nanos:89243000,user_nanos:88839000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:289287
peak_memory:0,elapsed_millis:85,input_rows_per_second:175980,output_rows_per_second:175980,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:85236738,cpu_nanos:85215000,user_nanos:85031000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291549
peak_memory:0,elapsed_millis:97,input_rows_per_second:155423,output_rows_per_second:155423,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:96510469,cpu_nanos:95113000,user_nanos:94433000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:289153
peak_memory:0,elapsed_millis:117,input_rows_per_second:127737,output_rows_per_second:127737,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:117428706,cpu_nanos:114931000,user_nanos:113926000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291395
peak_memory:0,elapsed_millis:91,input_rows_per_second:165372,output_rows_per_second:165372,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:90704106,cpu_nanos:90219000,user_nanos:89751000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291261
peak_memory:0,elapsed_millis:106,input_rows_per_second:141434,output_rows_per_second:141434,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:106056463,cpu_nanos:105823000,user_nanos:105200000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:290577
peak_memory:0,elapsed_millis:94,input_rows_per_second:158907,output_rows_per_second:158907,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:94394327,cpu_nanos:94142000,user_nanos:93697000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:289846
peak_memory:0,elapsed_millis:125,input_rows_per_second:119541,output_rows_per_second:119541,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:125479814,cpu_nanos:123750000,user_nanos:111631000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291722
peak_memory:0,elapsed_millis:162,input_rows_per_second:92849,output_rows_per_second:92849,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:161552397,cpu_nanos:160613000,user_nanos:127760000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:289892
peak_memory:0,elapsed_millis:107,input_rows_per_second:140792,output_rows_per_second:140792,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:106539760,cpu_nanos:101371000,user_nanos:98994000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:291229
peak_memory:0,elapsed_millis:88,input_rows_per_second:169527,output_rows_per_second:169527,input_megabytes:0,input_megabytes_per_second:0,wall_nanos:88481364,cpu_nanos:88485000,user_nanos:88386000,input_rows:15000,input_bytes:0,output_rows:15000,output_bytes:288919
                          sql_trail ::  107.978 cpu ms ::    0B peak memory :: in   15K,      0B,    139K/s,      0B/s :: out   15K,   284KB,    139K/s,  2.57MB/s